### PR TITLE
Type annotations

### PR DESCRIPTION
**Pull request by andrejbauer**
This PR does several things:

1. It was necessary to get rid of the polymorphic syntax and split it into two intermediate languages:

     * `Dsyntax` is desugared syntax
     * `Rsyntax` is runtime syntax

   The reason is that we need *two* kinds of annotations: `let`-bindings are annotated differently
   from arguments to functions.

   Even if there was some way to keep a single syntax, in the long run it would be impossible to live with
   it. It was really ugly.

2. ML-type annotations are indicated with `:>` on `let`-bindings and as arguments to functions:

    * `let f (x :> t1) (y :> t2) = ...`
    * `let g x y :> sch = ...`
    * `fun (x :> t1) (y :> t2) => ...`

   In the second form `t` is the type of the resulting function (this is different from  OCaml).

3. We allow TT-level annotations of the form `let x : t = c` which is equivalent to `let x = (c : t)`. While we might make sense of `let f x y : t = c`, we should not. Just keep it simple and make people write λ when they mean λ.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
> 	7f7468f
> 	Arguments in recursive definitions cannot be polymorphic.
>Because they are just like arguments of functions, which cannot
be polymorphic either (why?).

Do you mean why can't we have `fun (x :> forall a, a -> a) => foo`?

**Comment by andrejbauer** [comment]
I asked @christopherastone about this. It's not undoable, and Haskell has higher rank functions, but we lose principal types (not a big deal), and in general we bring in trickiness that does not seem to be warranted.

**Comment by SkySkimmer** [comment]
Ocaml does this with polymorphic record fields which put enough information in the syntax to make things nicer I think.

File: src/typing/typecheck.ml [hunk 1]
```diff
@@ -562,76 +597,84 @@ let add_tydef env (t, (params, def)) =
...
-    let env, (c, t) = Tyenv.at_toplevel env (comp c) in
-    let s = Mlty.ungeneralized_schema t in
-    let env = Tyenv.topadd_let x s env in
-    env, locate ~loc (Syntax.TopDynamic (x, s, c))
+    env, locate ~loc (Rsyntax.TopLetRec xycs)
+
+  | Dsyntax.TopDynamic (x, annot, c) ->
+    let env, (c, t) =
+      Tyenv.at_toplevel env
+        (comp c >>= fun (c, t) ->
```
Comment (SkySkimmer) [inline src/typing/typecheck.ml]: I sort of feel like this should be its own function, like check_annotated_comp or something, but I'm not sure.
If this comment inspires you please make a followup PR.

Comment (andrejbauer) [inline src/typing/typecheck.ml]: I take it this is now obsolete in view of PR #375 ?

Comment (SkySkimmer) [inline src/typing/typecheck.ml]: No.

**Pull request merged**